### PR TITLE
add mp3 encoder

### LINKED ISSUE #15: mp3 encoder?
First of all, thanks for building this great polyfill, it's a life saver on safari.

I saw there's a wav encoder built-in, wondering is there any chance for a mp3 encoder, maybe using [lamejs](https://github.com/zhuker/lamejs), so that recorded audio files could be played across chrome/firefox/safari?

### PR DESCRIPTION
**Pull request by youkaclub**
close #15

### CODE REVIEW

File: mpeg-encoder.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+module.exports = function () {
```
Comment (ai) [inline mpeg-encoder.js]: Do we need docs on how to use it?

File: index.js [hunk 2]
```diff
@@ -242,14 +242,14 @@ MediaRecorder.prototype = {
...
+ * This polyfill supports `audio/wav` and `audio/mpeg`.
  *
  * @param {string} mimeType The mimeType to check.
  *
- * @return {boolean} `true` on `audio/wav` MIME type.
+ * @return {boolean} `true` on `audio/wav` and `audio/mpeg` MIME type.
  */
 MediaRecorder.isTypeSupported = function isTypeSupported (mimeType) {
-  return /audio\/wave?/.test(mimeType)
+  return /audio\/(wave?|mpeg)/.test(mimeType)
```
Comment (ai) [inline index.js]: Does Polyfill support `audio/mpeg` out of the box? It will be better to enable MP3 encoding only when the user manually changed the coder (because not everyone need MP3, but the MP3 decoder takes a lot of client-side bundle).

File: mpeg-encoder.js (hunk 1 above)
Comment (youkaclub) [reply mpeg-encoder.js]: already documented https://github.com/ai/audio-recorder-polyfill#custom-encoder.
do you want to add mp3 example?

File: index.js (hunk 2 above)
Comment (youkaclub) [reply index.js]: the lamejs lib is loaded on demand https://github.com/ai/audio-recorder-polyfill/pull/43/files#diff-683fe31c51c17e6e3c61b9d26c0dd8eeR3

Comment (ai) [reply index.js]: Yeap, we need to add:

1. That we have out-of-the-box MP3 support
2. Guide with exact steps to enable MP3

File: mpeg-encoder.js (hunk 1 above)
Comment (ai) [reply mpeg-encoder.js]: Yeap, we need to add:

* That we have out-of-the-box MP3 support
* Guide with exact steps to enable MP3

File: index.js (hunk 2 above)
Comment (ai) [reply index.js]: Am I right, that `mpeg-encoder.js` file is not loaded to `index.js`? The developer needs to explicitly load it. So if the developer didn’t load it, Polyfill will not have MP3 support

File: mpeg-encoder.js (hunk 1 above)
Comment (youkaclub) [reply mpeg-encoder.js]: anyway, I've updated the readme with example how to use the new mpeg encoder.
https://github.com/ai/audio-recorder-polyfill/pull/43/commits/ad87bdb02489579a824717ce8ff6e9407cd489b2

Comment (ai) [reply mpeg-encoder.js]: Great. I added a few notes to polish the docs

File: index.js (hunk 2 above)
Comment (youkaclub) [reply index.js]: exactly, just like the documented example.
so maybe the isTypeSupported function should returns something like that:
`return MediaRecorder.prototype.mimeType === mimeType`

**Comment by youkaclub** [comment]
> Why do we need this change?
> https://github.com/ai/audio-recorder-polyfill/blob/master/package.json#L79

yes

File: mpeg-encoder.js (hunk 1 above)
Comment (youkaclub) [reply mpeg-encoder.js]: I updated the readme https://github.com/ai/audio-recorder-polyfill/pull/43/files#diff-04c6e90faac2675aa89e2176d2eec7d8R144

File: README.md [hunk 3]
```diff
@@ -164,8 +173,8 @@ you can change polyfill’s encoder:
 
 ```diff
   window.MediaRecorder = require('audio-recorder-polyfill')
-+ MediaRecorder.encoder = require('./ogg-opus-encoder')
-+ MediaRecorder.mimeType = 'audio/ogg'
++ MediaRecorder.encoder = require('audio-recorder-polyfill/ogg-opus-encoder')
```
Comment (ai) [inline README.md]: We do not have `audio-recorder-polyfill/ogg-opus-encoder`, let’s use old value

File: README.md [hunk 4]
```diff
@@ -144,6 +141,18 @@ recorder.addEventListener('dataavailable', e => {
 })
 ```
 
+### wav
+As default ,this polyfill saves records to `.wav` files. Compression
+is not very good, but encoding is fast and simple.
+
+### mp3
```
Comment (ai) [inline README.md]: The official way to write a name is `MP3` and `WAV`
https://en.wikipedia.org/wiki/MP3

File: index.js (hunk 2 above)
Comment (ai) [reply index.js]: I like `MediaRecorder.prototype.mimeType === mimeType` way

File: README.md [hunk 5]
```diff
@@ -9,7 +9,7 @@ Try it in **[online demo]** and see **[API]**.
 
 * **Spec compatible.** In the future when all browsers will support
   MediaRecorder, you will remove polyfill.
-* **Small.** 1.1 KB (minified and gzipped). No dependencies.
+* **Small.** 1.2 KB (minified and gzipped). No dependencies.
   It uses [Size Limit] to control size.
 * **One file.** In contrast to other recorders, this polyfill uses
   “inline worker” and don’t need a separated file for Web Worker.
```
Comment (ai) [inline README.md]: Let’s add MP3 support as extra feature

Comment (youkaclub) [reply README.md]: what do you mean?

File: mpeg-encoder.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      var fr = new FileReader()
+      fr.onload = function () {
+        resolve(fr.result)
+      }
+      // eslint-disable-next-line
+      fr.readAsArrayBuffer(this);
+      // eslint-disable-next-line
+    }.bind(this))
+  }
+  Blob.prototype.arrayBuffer = Blob.prototype.arrayBuffer || arrayBufferPolyfill
```
Comment (ai) [inline mpeg-encoder.js]: It is bad practice to modify the standard library.

If we need `ArrayBuffer` polyfill just for IE, we can ask developers to add it manually. Let’s avoid unnecessary polyfills for rarely used browsers.

File: mpeg-encoder.js (hunk 1 above)
Comment (ai) [inline mpeg-encoder.js]: Unfortunately, we need tests for this file as well.

You can check my test for WAV encoder https://github.com/ai/audio-recorder-polyfill/blob/master/test/wave-encoder.test.js

File: mpeg-encoder.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  var CHANNELS = 1
+  var KBPS = 128
+  var SAMPLE_RATE = 44100
+
+  // eslint-disable-next-line
+  var encoder = new lamejs.Mp3Encoder(CHANNELS, SAMPLE_RATE, KBPS)
```
Comment (ai) [inline mpeg-encoder.js]: What ESLint warning do we have here? Maybe I can recommend way to avoid disabling the linter.

File: README.md (hunk 5 above)
Comment (ai) [reply README.md]: I mean something like:

```
* **MP3** and **WAV** encoder support.
```

File: mpeg-encoder.js (hunk 7 above)
Comment (youkaclub) [reply mpeg-encoder.js]: undef lamejs

File: mpeg-encoder.js (hunk 1 above)
Comment (youkaclub) [reply mpeg-encoder.js]: it's really tough to test web worker with importScripts :(

File: mpeg-encoder.js (hunk 6 above)
Comment (youkaclub) [reply mpeg-encoder.js]: safari needs it as well and it's only few lines.

File: README.md (hunk 3 above)
Comment (youkaclub) [reply README.md]: actually, this is the old old value :) but fine

File: mpeg-encoder.js (hunk 7 above)
Comment (ai) [reply mpeg-encoder.js]: You can use `window.lamejs`. It will not have warning.

File: mpeg-encoder.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
+module.exports = function () {
+  // eslint-disable-next-line
+  importScripts('https://cdnjs.cloudflare.com/ajax/libs/lamejs/1.2.0/lame.min.js')
```
Comment (ai) [inline mpeg-encoder.js]: Can we use it from npm?

File: README.md (hunk 3 above)
Comment (ai) [reply README.md]: Old value was `./ogg-opus-encoder` where `.` is a developer’s project

File: mpeg-encoder.js (hunk 1 above)
Comment (ai) [reply mpeg-encoder.js]: Another reason to use `lamejs` from npm

File: mpeg-encoder.js (hunk 6 above)
Comment (ai) [reply mpeg-encoder.js]: OK, let’s use:

```js
var blobToArray
if (Blob.prototype.arrayBuffer) {
  blobToArray = function (blob) {
    return Blob.prototype.arrayBuffer
  }
} else {
  blobToArray = function (blob) {
    // polyfill
  }
}
```

In this case, we will not change the browser system classes

File: README.md (hunk 5 above)
Comment (youkaclub) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (youkaclub) [reply README.md]: done

File: index.js (hunk 2 above)
Comment (youkaclub) [reply index.js]: done

File: yarn.lock [hunk 9]
```diff
@@ -1,12538 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ai) [inline yarn.lock]: Seems like you accidentally delete `yarn.lock`

File: test/demo/index.js [hunk 10]
```diff
@@ -1,5 +1,5 @@
 var EVENTS = ['start', 'stop', 'pause', 'resume']
-var TYPES = ['audio/webm', 'audio/ogg', 'audio/wav']
+var TYPES = ['audio/webm', 'audio/ogg', 'audio/wav', 'audio/mpeg']
```
Comment (ai) [inline test/demo/index.js]: I think we need to revert this change since we reverted `isTypeSupported`

File: mpeg-encoder.js (hunk 8 above)
Comment (youkaclub) [reply mpeg-encoder.js]: no without change the current way to create a new worker (pass the function string). you can't use require without bundle it first.

File: mpeg-encoder.js (hunk 1 above)
Comment (ai) [reply mpeg-encoder.js]: OK, another way to test encoder is to emulate `window.importScripts = () => true`

File: README.md (hunk 3 above)
Comment (youkaclub) [reply README.md]: OCD FTW :)

File: mpeg-encoder.js (hunk 7 above)
Comment (youkaclub) [reply mpeg-encoder.js]: done

File: mpeg-encoder.js (hunk 6 above)
Comment (youkaclub) [reply mpeg-encoder.js]: done

File: yarn.lock (hunk 9 above)
Comment (youkaclub) [reply yarn.lock]: sorry. reverted

File: test/index.test.js [hunk 11]
```diff
@@ -42,7 +42,6 @@ beforeEach(() => {
 
 it('checks audio format support', () => {
   expect(MediaRecorder.isTypeSupported('audio/wav')).toBeTruthy()
-  expect(MediaRecorder.isTypeSupported('audio/wave')).toBeTruthy()
```
Comment (ai) [inline test/index.test.js]: We still need this test

File: test/demo/index.js (hunk 10 above)
Comment (youkaclub) [reply test/demo/index.js]: done

File: test/index.test.js (hunk 11 above)
Comment (youkaclub) [reply test/index.test.js]: why one is not enough?
unless you check all of them (audio/wave,audio/wav,audio/x-wav,audio/x-pn-wav)
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

Comment (youkaclub) [reply test/index.test.js]: I updated the test btw

Comment (ai) [reply test/index.test.js]: It will force us to release major update since we broke the previous API.

But OK, it is a good pint that there are many more MIME formats.

File: mpeg-encoder.js (hunk 1 above)
Comment (youkaclub) [reply mpeg-encoder.js]: do you prefer to add build step for encoders?

Comment (ai) [reply mpeg-encoder.js]: Nope. It is OK to take `lamejs` from CDN.

I told you about the test for `mpeg-encoder`. You can mock `importScript` and insert fake `window.lamejs` in the test.

Comment (youkaclub) [reply mpeg-encoder.js]: done

File: test/mpeg-encoder.test.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
... 4, 76, 65, 77, 69, 3, 0, 9, 8, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
```
Comment (ai) [inline test/mpeg-encoder.test.js]: You can’t just disable line because of the length.

Split it to multiple lines:

```js
expect(messages).toEqual([[
  255, 251, 144, 196, 0, 0, 21, 242, 11, 96, 128, 140, 
  211, 194, 200, 188, 108, 116, 51, 27, 113, 78
]])
```

File: mpeg-encoder.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  function encode (buffer) {
+    for (var i = 0; i < buffer.length; i++) {
+      buffer[i] = buffer[i] * 32767.5
+    }
+
+    var buf = encoder.encodeBuffer(buffer)
+    if (buf.length > 0) {
+      recorded = concat(recorded, buf)
```
Comment (ai) [inline mpeg-encoder.js]: CI falls because the test doesn’t cover this line.

(Yeap, it will be the last thing =^_^=)

Comment (youkaclub) [reply mpeg-encoder.js]: done :)

File: test/mpeg-encoder.test.js (hunk 12 above)
Comment (youkaclub) [reply test/mpeg-encoder.test.js]: done

**Comment by ai** [comment]
Everything looks good. I will merge and release tomorrow.

Sorry, hard day today.

**Comment by youkaclub** [comment]
> Everything looks good. I will merge and release tomorrow.
> 
> Sorry, hard day today.

no problem, thanks!

**Comment by ai** [comment]
Thanks for the work.

Released 0.2 version.
